### PR TITLE
feat: implement assembler

### LINKED ISSUE #19: Implement all the modules in `Assembler` abstraction
We want an `assembler` that translates programs written in `Hack assembly language` into the binary code understood by the `Hack hardware platform`

The modules we want to implement
- `Parser` (encapsulates access to input codes)
- `Code` (translates hack assembly mnemonics into binary codes)
- `Symbol Table` (keeps a correspondence between symbolic labels and numeric addresses)

### PR DESCRIPTION
**Pull request by henok-tesfaye**
Closes #19 

**What kind of change does this PR introduce?**
It is a feature, which implements the assembler abstraction. 

**Does this PR introduce a breaking change?**
No

**What needs to be documented once your changes are merged?**
Nothing

### CODE REVIEW

File: src/abstractions/software/assembler/index.test.js [hunk 1]
```diff
@@ -1,21 +1,110 @@
 import Assembler from './index'
-import { symbolLess } from './mockData'
+import { symbolLess, symbol } from './mockData'
+import { getPredefinedSymbols } from './symbolTable/utils'
 
 describe('Assember class', () => {
-  it('should translate add assembly code to machine code', () => {
+  it('should translate add symbol less assembly code to machine code', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/index.test.js:6]:
```suggestion
  it('should translate the ADD symbol less assembly code to machine code', () => {
```

File: src/abstractions/software/assembler/index.test.js [hunk 2]
```diff
@@ -1,21 +1,110 @@
...
     const assembler = new Assembler(symbolLess.ADD_ASSEMBLY_CODE)
     expect(assembler.assemble()).toBe(symbolLess.ADD_MACHINE_CODE.trim())
   })
-  it('should translate max assembly code to machine code', () => {
+  it('should translate max symbol less assembly code to machine code', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/index.test.js:10]:
```suggestion
  it('should translate the MAX symbol less assembly code to machine code', () => {
```

File: src/abstractions/software/assembler/index.test.js [hunk 3]
```diff
@@ -1,21 +1,110 @@
...
     const assembler = new Assembler(symbolLess.MAX_ASSEMBLY_CODE)
     expect(assembler.assemble()).toBe(symbolLess.MAX_MACHINE_CODE.trim())
   })
-  it('should translate rect assembly code to machine code', () => {
+  it('should translate rect symbol less assembly code to machine code', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/index.test.js:14]:
```suggestion
  it('should translate the RECT symbol less assembly code to machine code', () => {
```

File: src/abstractions/software/assembler/index.test.js [hunk 4]
```diff
@@ -1,21 +1,110 @@
...
     const assembler = new Assembler(symbolLess.RECT_ASSEMBLY_CODE)
     expect(assembler.assemble()).toBe(symbolLess.RECT_MACHINE_CODE.trim())
   })
-  it('should translate pong assembly code to machine code', () => {
+  it('should translate pong symbol less assembly code to machine code', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/index.test.js:18]:
```suggestion
  it('should translate the PONG symbol less assembly code to machine code', () => {
```

File: src/abstractions/software/assembler/index.test.js [hunk 5]
```diff
@@ -1,21 +1,110 @@
...
     const assembler = new Assembler(symbolLess.PONG_ASSEMBLY_CODE)
     expect(assembler.assemble()).toBe(symbolLess.PONG_MACHINE_CODE.trim())
   })
+
+  it('should translate max symbol assembly code to machine code', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/index.test.js:23]:
```suggestion
  it('should translate the MAX assembly code to machine code', () => {
```

File: src/abstractions/software/assembler/index.test.js [hunk 6]
```diff
@@ -1,21 +1,110 @@
...
+    const assembler = new Assembler(symbol.MAX_ASSEMBLY_CODE)
+    expect(assembler.assemble()).toBe(symbol.MAX_MACHINE_CODE.trim())
+  })
+
+  it('should translate rect symbol assembly code to machine code', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/index.test.js:28]:
```suggestion
  it('should translate the RECT assembly code to machine code', () => {
```

File: src/abstractions/software/assembler/index.test.js [hunk 7]
```diff
@@ -1,21 +1,110 @@
...
+    const assembler = new Assembler(symbol.RECT_ASSEMBLY_CODE)
+    expect(assembler.assemble()).toBe(symbol.RECT_MACHINE_CODE.trim())
+  })
+
+  it('should translate pong symbol assembly code to machine code', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/index.test.js:33]:
```suggestion
  it('should translate the PONG assembly code to machine code', () => {
```

File: src/abstractions/software/assembler/index.test.js [hunk 8]
```diff
@@ -1,21 +1,110 @@
...
+    const assembler = new Assembler(symbol.PONG_ASSEMBLY_CODE)
+    expect(assembler.assemble()).toBe(symbol.PONG_MACHINE_CODE.trim())
+  })
+
+  it('should bound ROM address to labels, when label is the first instruction', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/index.test.js:38]:
```suggestion
  it('should bind ROM address to a label when label is the first instruction', () => {
```

File: src/abstractions/software/assembler/index.test.js [hunk 9]
```diff
@@ -1,21 +1,110 @@
...
+    const assembler = new Assembler(getFirstLabelAssembleTestCode())
+    const expectedSymbolTable = { ...getPredefinedSymbols(), LOOP: 0 }
+    assembler.firstPass()
+    const actualSymbolTable = assembler.getSymbolTable()
+    expect(actualSymbolTable).toEqual(expectedSymbolTable)
+  })
+
+  it('should bound ROM address to labels, when label is the last instruction', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/index.test.js:46]:
```suggestion
  it('should bind ROM address to a label when label is the last instruction', () => {
```

File: src/abstractions/software/assembler/index.test.js [hunk 10]
```diff
@@ -1,21 +1,110 @@
...
+    const assembler = new Assembler(getLastLabelAssembleTestCode())
+    const expectedSymbolTable = { ...getPredefinedSymbols(), LOOP: 2 }
+    assembler.firstPass()
+    const actualSymbolTable = assembler.getSymbolTable()
+    expect(actualSymbolTable).toEqual(expectedSymbolTable)
+  })
+
+  it('should bound ROM address to labels', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/index.test.js:54]:
```suggestion
  it('should bind ROM addresses to labels', () => {
```

File: src/abstractions/software/assembler/index.test.js [hunk 11]
```diff
@@ -1,21 +1,110 @@
...
+    const assembler = new Assembler(getAssembleTestCode())
+    const expectedSymbolTable = { ...getPredefinedSymbols(), LOOP: 1, END: 3, EXIT: 5 }
+    assembler.firstPass()
+    const actualSymbolTable = assembler.getSymbolTable()
+    expect(actualSymbolTable).toEqual(expectedSymbolTable)
+  })
 })
+
+const getFirstLabelAssembleTestCode = () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/index.test.js:63]:
```suggestion
const getLabelFirstAssemblyTestCode = () => {
```

File: src/abstractions/software/assembler/index.test.js [hunk 12]
```diff
@@ -1,21 +1,110 @@
...
+    // File name: projects/06/pong/PongL.asm
+
+    // Symbol-less version of the Pong.asm program.
+    (LOOP)
+    @256
+    D=A
+  `
+}
+
+const getLastLabelAssembleTestCode = () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/index.test.js:77]:
```suggestion
const getLabelLastAssemblyTestCode = () => {
```

File: src/abstractions/software/assembler/index.test.js [hunk 13]
```diff
@@ -1,21 +1,110 @@
...
+
+    // Symbol-less version of the Pong.asm program.
+
+    @256
+    D=A
+    (LOOP)
+  `
+}
+
+const getAssembleTestCode = () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/index.test.js:92]:
```suggestion
const getAssemblyTestCode = () => {
```

File: src/abstractions/software/assembler/mockData/index.js [hunk 14]
```diff
@@ -1,3 +1,4 @@
 import symbolLess from './symbolLess'
```
Comment (mezzzi) [inline src/abstractions/software/assembler/mockData/index.js:1]: also change references in the test file
```suggestion
import symbolLessAssembly from './symbolLessAssembly'
```

File: src/abstractions/software/assembler/mockData/index.js [hunk 15]
```diff
@@ -1,3 +1,4 @@
...
+import symbol from './symbol'
```
Comment (mezzzi) [inline src/abstractions/software/assembler/mockData/index.js:2]: I think `symbol` is confusing, let's just have `assembly` and `symbolLessAssembly`
Also for this change, update references in the test file
```suggestion
import assembly from './assembly'
```

File: src/abstractions/software/assembler/symbolTable/index.js [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+  constructor () {
+    this.table = {
+      ...getPredefinedSymbols()
+    }
+  }
+
+  /**
+   * @param {string} symbol symbol
+   * @param {number} address address
+   * Add  (symbol, address) pair to the table
```
Comment (mezzzi) [inline src/abstractions/software/assembler/symbolTable/index.js:16]: The function description should be above the param descriptions, right?
```suggestion
   * Add  (symbol, address) pair to the table
   * @param {string} symbol symbol
   * @param {number} address address of the symbol
```

File: src/abstractions/software/assembler/symbolTable/index.js [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+   */
+  addEntry (symbol, address) {
+    if (this.contains(symbol)) return
+    if (symbol && (address || address === 0)) this.table[symbol] = address
```
Comment (mezzzi) [inline src/abstractions/software/assembler/symbolTable/index.js:20]:
```suggestion
    if (symbol && (address || address === 0)) this.table[symbol] = address
    // what if the above conditions are not met?
    // returning an error with a helpful message will make debugging easier
     return new Error(
       `Error in symbolTable.addEntry. For symbol=${symbol} and address=${address}`)    
```

File: src/abstractions/software/assembler/symbolTable/index.test.js [hunk 18]
```diff
@@ -0,0 +1,32 @@
+import SymbolTable from './index'
+
+describe('SymbolTable class', () => {
+  const symbolTable = new SymbolTable()
+
+  it('should return predifined symbols', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/symbolTable/index.test.js:6]:
```suggestion
  it('should return predefined symbols', () => {
```

File: src/abstractions/software/assembler/symbolTable/index.test.js [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+    expect(symbolTable.getAdress('R0')).toBe(0)
+    expect(symbolTable.getAdress('SCREEN')).toBe(16384)
+  })
+
+  it('should add entry (symbol, address) to the table and return the associated address', () => {
+    symbolTable.addEntry('END', 0)
+    expect(symbolTable.getAdress('END')).toBe(0)
+  })
+
+  it('shoudn\'t add (symbol, address) pair, if symbol exist', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/symbolTable/index.test.js:17]:
```suggestion
  it('shoudn\'t add (symbol, address) pair, if symbol exists', () => {
```

File: src/abstractions/software/assembler/symbolTable/index.test.js [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+    symbolTable.addEntry('age', 23)
+    symbolTable.addEntry('age', 25)
+    expect(symbolTable.getAdress('age')).toBe(23)
+  })
+
+  it('return true if symbol exist', () => {
```
Comment (mezzzi) [inline src/abstractions/software/assembler/symbolTable/index.test.js:23]:
```suggestion
  it('return true if symbol exists', () => {
```

File: src/abstractions/software/assembler/index.js [hunk 21]
```diff
@@ -1,37 +1,88 @@
...
+    this.symbolTable = new SymbolTable()
+    this.writer = new Writer()
+    this.RAMAddress = 16
   }
 
   /**
-   * @returns {string} corresponding machine code for the given assembly code
+   * It marches through the program and bound memory address to the labels,
+   * in the assebmly code
+   * @param {string} assemblyCode assebmly code
```
[the remark this answers is not in the document]
Comment (henok-tesfaye) [reply src/abstractions/software/assembler/index.js:18]: yes, thank you!

File: src/abstractions/software/assembler/index.js [hunk 22]
```diff
@@ -1,37 +1,88 @@
...
           break
       }
       parser.advance()
     }
+  }
 
-    return writer.getResult()
+  /**
+   * It will bound variables to ram addresses
+   * @param {string} symbol symbol
```
[the remark this answers is not in the document]
Comment (henok-tesfaye) [reply src/abstractions/software/assembler/index.js:60]: `@return {number} the RAM address value of the given variable`
This function can also return `ROM` addresses for labels.

**Pull request merged**